src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs

```diff
@@ -0,0 +1,176 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.Unity.Analyzers.Resources;
+
+namespace Microsoft.Unity.Analyzers
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class ImproperSerializeFieldAnalyzer : DiagnosticAnalyzer
+	{
+		public const string Id = "UNT0013";
+
+		private static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+			Id,
+			title: Strings.ImproperSerializeFieldDiagnosticTitle,
+			messageFormat: Strings.ImproperSerializeFieldDiagnosticMessageFormat,
+			category: DiagnosticCategory.Correctness,
+			defaultSeverity: DiagnosticSeverity.Info,
+			isEnabledByDefault: true,
+			description: Strings.ImproperSerializeFieldDiagnosticDescription);
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			context.RegisterSyntaxNodeAction(AnalyzePropertyDeclaration, SyntaxKind.PropertyDeclaration);
+			context.RegisterSyntaxNodeAction(AnalyzeFieldDeclaration, SyntaxKind.FieldDeclaration);
+		}
+
+		private static void AnalyzePropertyDeclaration(SyntaxNodeAnalysisContext context)
+		{
+			var property = (MemberDeclarationSyntax)context.Node;
+			if (!(property is PropertyDeclarationSyntax))
+				return;
+
+			var symbol = context.SemanticModel.GetDeclaredSymbol(property);
+			if (symbol == null)
+				return;
+
+			if (!PropertyHasInvalidSerializeField(symbol, out var propertyName))
+				return;
+
+			context.ReportDiagnostic(Diagnostic.Create(Rule, property.GetLocation(), propertyName));
+		}
+
+		private static void AnalyzeFieldDeclaration(SyntaxNodeAnalysisContext context)
+		{
+			var field = (FieldDeclarationSyntax)context.Node;
+			if (!(field is FieldDeclarationSyntax))
+				return;
+
+			List<ISymbol> symbols = new List<ISymbol>();
+			foreach (var variable in field.Declaration.Variables)
+			{
+				var symbol = context.SemanticModel.GetDeclaredSymbol(variable);
+				if (symbol != null)
+					symbols.Add(symbol);
+			}
+
+			if (symbols.FirstOrDefault() == null)
+				return;
+
+			if (!FieldHasInvalidSerializeField(symbols, out var fieldName))
```

Comment:
This looks suspicious, I think we need a new test for that scenario. One FieldDeclaration can declare many field, we need to test the following scenario, which is valid:

```csharp
[SerializeField] private Transform transformA, transformB;
```

And the following scenario which is redundant:


```csharp
[SerializeField] public Transform transformA, transformB;
```

Reply:
C# doesn't let you add an attribute per field on the same declaration, so that helps.

Reply:
Yeah, that makes this pretty straightforward. I'll streamline this a bit and add new test cases.